generic: mtd: spinand: fix Foresee update_cache_variants initialization

dimfishr opened the pull request:
Fix update_cache_variants initialization for Foresee NAND with bad blocks

Tested on Xiaomi AX3000T + F35SQA001G with bad blocks and without bad blocks

See issue https://github.com/openwrt/openwrt/issues/17962

sc16me commented:
Tested-by: Aleksandr Danilov [sc16me@gmail.com](mailto:sc16me@gmail.com)

ServError commented:
Likely want the same changes in https://github.com/openwrt/openwrt/blob/main/package/boot/uboot-mediatek/patches/101-01-mtd-spinand-add-support-for-FORESEE-F35SQA002G.patch

Any idea if the slow read cache entry 'SPINAND_PAGE_READ_FROM_CACHE_OP(false, 0, 1, NULL, 0)' is problematic as well?

nstepanets commented:
A similar problem has already been discussed, perhaps it is better to use this patch?
https://github.com/openwrt/openwrt/pull/16915#issuecomment-2487184967

sc16me commented:
A two-line patch that completely solves the problem. If it were 1,000 lines long, it would have been accepted faster…

hauke commented on target/linux/generic/backport-6.6/411-v6.7-mtd-spinand-add-support-for-FORESEE-F35SQA002G.patch:
```diff
@@ -63,8 +63,8 @@ Link: https://lore.kernel.org/linux-mtd/20231002140458.147605-1-mmkurbanov@salut
 +static SPINAND_OP_VARIANTS(update_cache_variants,
-+		SPINAND_PROG_LOAD_X4(false, 0, NULL, 0),
-+		SPINAND_PROG_LOAD(false, 0, NULL, 0));
++		SPINAND_PROG_LOAD_X4(true, 0, NULL, 0),
++		SPINAND_PROG_LOAD(true, 0, NULL, 0));
```
Please do not modify this file, but add a separate patch with this change.

This patch is  a backport of this upstream commit: https://github.com/torvalds/linux/commit/f447318fb1d156b4b6da79266724c7ee347d1b59

When you modify this patch changes are high that we just drop this patch when we go to kernel 6.12. 

Please send this change also to the upstream Linux community.

dimfishr replied:
Done.  I've created standalone patch.

dimfishr commented:
Test build for AX3000T is available here:
https://github.com/dimfishr/openwrt/releases/tag/filogic-2025-02-13

ServError commented:
> Test build for AX3000T is available here: https://github.com/dimfishr/openwrt/releases/tag/filogic-2025-02-13

The u-boot in your test release still uses an unpatched driver.

dimfishr commented:
> The u-boot in your test release still uses an unpatched driver.

Thanx. I've added uboot patch as well.

dimfishr commented:
New build with uboot is here: https://github.com/dimfishr/openwrt/releases/tag/filogic-2025-02-17

csharper2005 commented:
I suggest to split the changes into 2 commits: `uboot-mediatek:` and `kernel:`. Backporting to the 24.10 branch in the future will only require a second commit since U-Boot 2025.01 isn't backported to the 24.10 branch (and most likely won't be).

hauke commented:
I do not understand why you have to change the `update_cache_variants` function from `Random Program Load Data` (`0x84`) to `Program Load Data` (`0x02`). Please explain this in the commit message. 

The datasheet defines both operations like this:
> The Program Load Data command has to be issued prior to Random Program Load Data command for
random page program. The difference is that Program Load Data command will reset the unused the
data bytes in the Data Buffer to FFh value, while Random Program Load Data command will only update
the data bytes that are specified by the command input sequence, the rest of the Data Buffer will remain
unchanged

https://www.longsys.com/uploads/LM-00006FORESEEF35SQA002GDatasheet_1650183701.pdf

dimfishr commented:
> I do not understand why you have to change the `update_cache_variants` function from `Random Program Load Data` (`0x84`) to `Program Load Data` (`0x02`). Please explain this in the commit message.

I used this explanation based on F35SQA001G datasheet:
https://github.com/openwrt/openwrt/pull/16915#issuecomment-2467696294

dimfishr commented:
> I suggest to split the changes into 2 commits

Sounds reasonable. I've split.

dmsza commented:
@dimfishr isn't your u-boot commit going to conflict with this u-boot patch below committed about 9 days ago?

https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/boot/uboot-mediatek/patches/101-01-mtd-spinand-add-support-for-FORESEE-F35SQA002G.patch;h=deeb37490591b0e73f021a30749498560fefed0e;hb=302c717aa3a434676a32a4dde8b10ec2c03798b5

nstepanets commented:
> A two-line patch that completely solves the problem. If it were 1,000 lines long, it would have been accepted faster…

Is it possible to make sure that the SPI controller will transmit at least 2048 bytes? If the SPI controller cannot transfer 2048 bytes in one step, 02H or 32H commands will result in data loss...

csharper2005 commented:
Is this related to the same topic?
https://github.com/torvalds/linux/commit/042087247835dad1ec5e39052abf022fd13c6326.patch

dimfishr commented:
> Is this related to the same topic? 

Yep, looks like this huge patchwork (27 patches) will automatically fix this problem. But for 6.6 kernel I don't think that is good decision ta backport 27 patches or even more related stuff.
For the new kernel this patch can be simply omitted.

csharper2005 commented:
> Yep, looks like this huge patchwork (27 patches) will automatically fix this problem. But for 6.6 kernel I don't think that is good decision to backport 27 patches or even more related stuff. For the new kernel this patch can be simply omitted.

@hauke asked to upstream your patch to the mainline linux ( https://github.com/openwrt/openwrt/pull/17963#discussion_r1955332163). If this is a strict condition, there are no other options but to upstream this patch or backport the huge patchwork.

hauke commented:
> > I do not understand why you have to change the `update_cache_variants` function from `Random Program Load Data` (`0x84`) to `Program Load Data` (`0x02`). Please explain this in the commit message.
> 
> I used this explanation based on F35SQA001G datasheet: [#16915 (comment)](https://github.com/openwrt/openwrt/pull/16915#issuecomment-2467696294)

When is the `update_cache_variants` used?
I do not understand why you have to reset the unused data bits in the duffer when reading a page from the flash into the cache. 

How did upstream Linux fix the problem? Is it only this commit https://github.com/torvalds/linux/commit/042087247835dad1ec5e39052abf022fd13c6326 ?

fildunsky commented:
Hope this useful PR will be approved and will be added at least to 24.10.2

hauke commented:
If this is a bug in the generic Linux bad block handling, I do not understand why you change something in the one flash driver. I still haven't understood why your change is fixing the problem. I am concerned that this change breaks some other functionality with the flash chip and that we have to maintain it forever because it does not get upstream. 

I am not a mtd expert, please contact the Linux upstream mtd mailing list. If there is a fix already in a more recent Linux upstream version please try to backport that instead.

I will not merge it as is, maybe someone else does.

dimfishr commented:
I've created new fix based on patchwork for 6.12 kernel.
New build has been testes on 2 Foresee devices wit Bad Blocks with stock and ubootmod versions
https://github.com/dimfishr/openwrt/releases/tag/filogic-2025-04-18

dimfishr commented:
> Thank you for the update. Do you know if the discussion on the patch continued?

No - I've just picked.

csharper2005 commented:
> How did upstream Linux fix the problem? Is it only this commit [torvalds/linux@0420872](https://github.com/torvalds/linux/commit/042087247835dad1ec5e39052abf022fd13c6326) ?

2 users of affected Xiaomi AX3000t devices reported that this patch helps. I've opened a PR - https://github.com/openwrt/openwrt/pull/18545.

dimfishr commented:
Excellent! I'm Closing this PR 
See new one https://github.com/openwrt/openwrt/pull/18545

fildunsky commented:
@Phantom2332 yes

geokvant commented:
Had same problem with AX3000T+AN5588+Foresee memory. 24.10.1 with patch included solved problem by UART recovery method
Thanks for solution @dimfishr (от души, хоть и пришлось научиться UART)

SkyWalker1726 commented:
> Had same problem with AX3000T+AN5588+Foresee memory. 24.10.1 with patch included solved problem by UART recovery method Thanks for solution @dimfishr (от души, хоть и пришлось научиться UART)

Where did you find it ? Is it on github ?

fildunsky commented:
> Where did you find it ? Is it on github ?

What do you mean? The patch is right here, just add it to your clean vanilla OpenWRT 24.10.1 build and that's it.

SkyWalker1726 commented:
> > Where did you find it ? Is it on github ?
> 
> What do you mean? The patch is right here in this PR, just add it to your clean vanilla OpenWRT 24.10.1 build and that's it.

I am not familiar with the process .
Would you mind explaining how to do that ?
Thanks

dimfishr commented:
@SkyWalker1726 Pull requests are used for developers and testers. 
For basic questions please use forum https://forum.openwrt.org/t/openwrt-support-for-xiaomi-ax3000t/180490
of official docs https://openwrt.org/docs/guide-developer/toolchain/use-buildsystem

csharper2005 commented:
> is it correct understanding that this is already fixed in kernel 6.12

Linux kernel (any version) is not fixed. So there's no acceptable patches for backporting.

csharper2005 commented:
It's safer use custom builds with this patch for devices with FORESEE chip.

dimfishr commented:
> Any thoughts on this?

What can I say?
My first proposal to always update cache only for specific Foresee memory wasn't accepted - maybe we can use it as small hack and drop immediately after new patch will be accepted?

csharper2005 commented:
1. The current version of the patch keep OpenWrt u-boot unpatched;
2. The pending patch was rejected by the linux kernel maintainer since the patch changes global wdesc template - https://patchwork.ozlabs.org/project/linux-mtd/patch/20241119093949.3014-1-SkyLake.Huang@mediatek.com/;
3. The bug is very annoying for the users and clean solution is absent at this moment.

So I suggest apply the previous version of the patches (https://github.com/openwrt/openwrt/commit/71f5c379df8a80745b38daa070ebdc53a5187b80, https://github.com/openwrt/openwrt/commit/02fadeeb5e584bc6f77fd4fc02b23f2dde270b77) as a workaround but move them to the `hack-6.6` and `hack-6.12` dirs (not `pending-`). These patches:
1. Have been successfully tested by hundreds of users (OpenWrt and [u-boot](https://github.com/csharper2005/bl-mt798x/blob/master/uboot-mtk-20230718-09eda825/drivers/mtd/nand/spi/foresee.c#L28)  both);
2. Are trivial and hence don't require a lot of resources for maintaining.

@981213 @hauke нow do you like this temporary solution? If it doesn't suite, what other options are there?

dimfishr commented:
@dmsza thanx for testing this on other devices. According to this new information I've moved previous patches to hack-* folders as @csharper2005 suggested.
@hauke since kernel patch is not compatible with other devices i don't think it will be ever merged, so maybe we can use hacks to help users that are suffering with UART only restore method.

New test build is available here: https://github.com/dimfishr/openwrt/releases/tag/filogic-2025-07-01

dimfishr committed:
generic: mtd: spinand: force update_cache_variants to use reset for Foresee NAND

Force update_cache_variantsvariants to use reset for Foresee NAND with bad blocks.

Tested on Xiaomi AX3000T + F35SQA001G with bad blocks and without bad blocks

Signed-off-by: Dim Fish <dimfish@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17963
Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 commented:
one cosmetic comment: The commit should be split into two, one for u-boot and the other for the kernel.
Looks good to me otherwise :)

dimfishr committed:
uboot-mediatek: force update_cache_variants to use reset for Foresee NAND

Force update_cache_variantsvariants to use reset for Foresee NAND with bad blocks.

Tested on Xiaomi AX3000T + F35SQA001G with bad blocks and without bad blocks

Signed-off-by: Dim Fish <dimfish@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17963
Signed-off-by: Chuanhong Guo <gch981213@gmail.com>
